### PR TITLE
fix: Fix handling of tags in awl.flatten()

### PR DESCRIPTION
**Pull request by dannygoldstein**
Internal JIRA: https://wandb.atlassian.net/browse/WB-14959

### Changes

1. **`weave/ops_arrow/list_.py`**: Updated `is_list_arrowweavelist` to use `assign_type` instead of a simple `isinstance` check. This fixes a bug where tagged lists were not being detected properly.
2. **`weave/ops_arrow/list_ops.py`**: Enhanced `flatten` method to handle tags, with specific code paths for tagged values. 
3. **`weave/tests/test_arrow.py`**: Added a new test `test_flatten_handles_tagged_lists` to validate that the `flatten` method correctly handles lists with tags.

### Context

In `list_ops.py`, the primary change is the enhancement of the `flatten` method. Previously, the method did not handle tagged values when flattening an ArrowWeaveList. The new implementation unwraps tags, processes them accordingly, and merges the flattened values with the tags if they exist.

The added test in `test_arrow.py` ensures that this new functionality works as expected. It generates a nested tagged list and confirms that, after flattening, the resultant ArrowWeaveList retains the correct tags and values.

### Impact

This change provides better support for ArrowWeaveLists with tags in the `flatten` method, ensuring that tags are preserved and correctly associated with their respective values after flattening.

### CODE REVIEW

File: weave/ops_arrow/list_ops.py [hunk 1]
```diff
@@ -888,10 +888,46 @@ def flatten_return_type(input_types):
...
+            # while this technically calls to_pylist(), it is not actually
+            # breaking any data out to python in our internal weave definition,
+            # because it is just a list of integers that are the list offsets
+            # that are passed to the pa.repeat() call below
+
+            num_repeats = pc.subtract(
+                values.offsets[1:], values.offsets[:-1]
+            ).to_pylist()
+
+            for tag, rnum in zip(tags, num_repeats):
```
Comment (shawnlewis) [inline weave/ops_arrow/list_ops.py:917]: Doesn't this end up doing what flatten does? I'm confused, might need a hint.

File: weave/ops_arrow/list_ops.py [hunk 2]
```diff
@@ -888,10 +888,46 @@ def flatten_return_type(input_types):
...
+                arr.tagged_value_value(),
+                arr.tagged_value_tag(),
+            )
+
+            values = value_awl._arrow_data
+            tags = tags_awl._arrow_data
+
+            tag_arrays = []
+
+            # while this technically calls to_pylist(), it is not actually
```
[the remark this answers is not in the document]
Comment (dannygoldstein) [reply weave/ops_arrow/list_ops.py:908]: No this preserves dict encoding

File: weave/ops_arrow/list_ops.py (hunk 1 above)
Comment (shawnlewis) [reply weave/ops_arrow/list_ops.py:917]: ok, well we still can't use to_pylist :). need to figure out how to do this in arrow. I can help here

Comment (dannygoldstein) [reply weave/ops_arrow/list_ops.py]: Updated to keep things in arrow following https://github.com/apache/arrow/issues/27923#issuecomment-1378008351

**Review by tssweeney — APPROVED** [review]

**Pull request merged**